ENIGMAsystem/SHELL/Universal_System/Serialization/serialization_fwd_decl.h

```diff
@@ -0,0 +1,252 @@
+
+template <typename T>
+inline auto internal_serialize(T &&value);
+
+template <typename T>
+inline T internal_deserialize(std::byte *iter);
+
+template <typename T>
+inline void enigma_serialize(const T &value, std::size_t &len, std::vector<std::byte> &bytes);
+
+template <typename T>
+inline void enigma_deserialize(T &value, std::byte *iter, std::size_t &len);
+
+inline void enigma_internal_deserialize_variant(variant &value, std::byte *iter, std::size_t &len);
+
+template <typename T>
+inline auto internal_serialize_into_fn(std::byte *iter, T *value);
+
+template <typename T>
+typename std::enable_if<std::is_same_v<std::string, std::decay_t<T>>>::type inline internal_serialize_into_fn(
+    std::byte *iter, T &&value);
+
+template <typename T>
+typename std::enable_if<std::is_same_v<bool, std::decay_t<T>>>::type inline internal_serialize_into_fn(std::byte *iter,
+                                                                                                       T &&value);
+
+template <typename T>
+typename std::enable_if<std::is_same_v<var, std::decay_t<T>>>::type inline internal_serialize_into_fn(std::byte *iter,
+                                                                                                      T &&value);
+
+template <typename T>
+typename std::enable_if<std::is_base_of_v<variant, std::decay_t<T>> &&
+                        !std::is_same_v<var, std::decay_t<T>>>::type inline internal_serialize_into_fn(std::byte *iter,
+                                                                                                       T &&value);
+
+template <typename T>
+typename std::enable_if<(std::is_integral_v<std::decay_t<T>> ||
+                         std::is_floating_point_v<std::decay_t<T>>)&&!std::is_same_v<std::decay_t<T>, bool>>::
+    type inline internal_serialize_into_fn(std::byte *iter, T &&value);
+
+// needs a better place
+template <typename T>
+struct is_std_vector : std::false_type {};
+
+template <typename T, typename Alloc>
+struct is_std_vector<std::vector<T, Alloc>> : std::true_type {};
+
+template <typename T>
+constexpr static inline bool is_std_vector_v = is_std_vector<T>::value;
+
+template <typename T>
+struct is_std_map : std::false_type {};
+
+template <typename Key, typename T, typename Compare, typename Allocator>
+struct is_std_map<std::map<Key, T, Compare, Allocator>> : std::true_type {};
+
+template <typename T>
+constexpr bool is_std_map_v = is_std_map<T>::value;
+
+template <typename T>
+struct is_std_complex : std::false_type {};
+
+template <typename T>
+struct is_std_complex<std::complex<T>> : std::true_type {};
+
+template <typename T>
+constexpr bool is_std_complex_v = is_std_complex<T>::value;
+
+template <typename T>
+struct complex_inner_type;
+
+template <typename T>
+struct complex_inner_type<std::complex<T>> {
+  using type = T;
+};
+
+template <typename T>
+using complex_inner_type_t = typename complex_inner_type<T>::type;
```

Comment:
Hmm, my plan for this file was to have just six lines of code: the six forward declarations for the functions that the user can implement. This way it is automatically documenting and we can just point users here to tell them what functions to implement. Its not required to forward-declare the templates which are specialized for the various types, because the user will get them anyways when `serialization.h` is included. For the type traits, I would suggest making a `type_traits.h` header file within this directory itself.

Reply:
Cool solution, I will do that.